api/libretimeapi/models.py

```diff
@@ -0,0 +1,663 @@
+
+    class Meta:
+        managed = False
+        db_table = 'cc_stream_setting'
+
+GUEST = 'G'
+DJ = 'H'
+PROGRAM_MANAGER = 'P'
+ADMIN = 'A'
+
+USER_TYPE_CHOICES = (
+    (GUEST, 'Guest'),
+    (DJ, 'DJ'),
+    (PROGRAM_MANAGER, 'Program Manager'),
+    (ADMIN, 'Admin'),
+)
+
+class UserManager(BaseUserManager):
+    def create_user(self, username, type, email, first_name, last_name, password):
+        user = self.model(username=username,
+                          type=type,
+                          email=email,
+                          first_name=first_name,
+                          last_name=last_name)
+        user.set_password(password)
+        user.save(using=self._db)
+        return user
+
+    def create_superuser(self, username, email, first_name, last_name, password):
+        user = self.create_user(username, ADMIN, email, first_name, last_name, password)
+        return user
+
+    def get_by_natural_key(self, username):
+        return self.get(username=username)
+
+class User(AbstractBaseUser):
+    username = models.CharField(db_column='login', unique=True, max_length=255)
+    password = models.CharField(db_column='pass', max_length=255)  # Field renamed because it was a Python reserved word.
+    type = models.CharField(max_length=1, choices=USER_TYPE_CHOICES)
+    first_name = models.CharField(max_length=255)
+    last_name = models.CharField(max_length=255)
+    last_login = models.DateTimeField(db_column='lastlogin', blank=True, null=True)
+    lastfail = models.DateTimeField(blank=True, null=True)
+    skype_contact = models.CharField(max_length=1024, blank=True, null=True)
+    jabber_contact = models.CharField(max_length=1024, blank=True, null=True)
+    email = models.CharField(max_length=1024, blank=True, null=True)
+    cell_phone = models.CharField(max_length=1024, blank=True, null=True)
+    login_attempts = models.IntegerField(blank=True, null=True)
+
+    USERNAME_FIELD = 'username'
+    EMAIL_FIELD = 'email'
+    REQUIRED_FIELDS = ['type', 'email', 'first_name', 'last_name']
+    objects = UserManager()
+
+    def get_full_name(self):
+        return f'{self.first_name} {self.last_name}'
+
+    def get_short_name(self):
+        return f'{self.first_name}'
+
+    def set_password(self, password):
+        if not password:
+            self.set_unusable_password()
+        else:
+            self.password = hashlib.md5(password.encode()).hexdigest()
+
+    def is_superuser(self):
+        return self.type == ADMIN
+
+    def is_staff(self):
+        return self.type == ADMIN
+
+    def check_password(self, password):
+        if self.has_usable_password():
+            test_password = hashlib.md5(password.encode()).hexdigest()
+            return test_password == self.password
+        return False
+
+    def has_perms(self, perm_list, obj=None):
+        if not self.is_active:
+            return False
+        if self.is_superuser():
+            return True
+        if len(perm_list) == 0:
```

Comment:
Mainly because Permissions aren't yet implemented, so returning False would break things for everyone who is not an admin

Reply:
Gotcha... Giving everyone admin doesn't feel right even if it's only via the api.

In my case (FreeIPA) it won't matter because the users don't know the randomly-generated password that is in cc_subj.

Maybe we can hide bits that read users from cc_subj behind a feature flag (but keep the api-key implemented).

I have a feeling that porting the permissions to python might be a bit of an effort.

Reply:
The user accesss should be hidden behind the requirement that the auth user is either an admin or the current user, but I can't remember if that got lost during the work after that or not